cpp/src/arrow/compute/kernels/codegen_internal.h

```diff
@@ -407,12 +407,19 @@ ArrayKernelExec MakeFlippedBinaryExec(ArrayKernelExec exec);
 // Helpers for iterating over common DataType instances for adding kernels to
 // functions
 
+ARROW_EXPORT
 const std::vector<std::shared_ptr<DataType>>& BaseBinaryTypes();
```

Comment:
Hmm... this seems to imply that `codegen_internal.h` is included in tests, but that's a rather costly inclusion. Instead, we already have some declarations in `gtest_util.h` (see `BinaryArrowTypes` and friends).

Reply:
Those are for compile-time parameterization while the test in hash_aggregate_test.cc uses these for runtime parameterization. But it's not such a big deal, so I'll convert it to use TYPED_TEST.

Reply:
Well, these helpers would be globally useful, so perhaps put them at the end of `type.h`?

Reply:
I moved the helpers and also culled an underutilized one (DecimalTypeIds). AllTimeUnits was promoted to TimeUnit::values() instead.